### ISSUE TITLE
Compile with stable Rust

### ISSUE THREAD

**Issue by bstrie** [enarx/enarx#1372]
Make enarx and all dependencies compile on a stable release of Rust, using only stable Rust features. This may involve several approaches:

1. Remove the usage of any unstable features that are not necessary for functioning.
2. Pursue upstream stabilization for any unstable features that are necessary.
3. In cases where unstable feature usage is platform-specific, use conditional compilation to limit the need for a nightly compiler to only certain platforms.
4. Offer our own feature flags to allow users to opt-in to the use of nightly features.

# Feature status

- [ ] [`-Z bindeps`](https://github.com/rust-lang/cargo/issues/9096) (Cargo feature): high risk. This feature (a.k.a. "artifact dependencies") represents a substantial amount of new code in Cargo and a handful of issues have already surfaced: https://github.com/rust-lang/cargo/issues?q=is%3Aissue+is%3Aopen+artifact+label%3AZ-bindeps . Silver lining is that the fact that others are finding issues means that other people are eager to use this in earnest, which is important for eventual stabilization. The Cargo team may be reluctant to stabilize as long as outstanding issues exist. Furthermore, the Cargo team is relatively understaffed and may not have the energy necessary to pursue stabilization without significant external assistance.
- [x] [`-Z sparse-registry`](https://github.com/rust-lang/cargo/issues/9069) (Cargo feature): ~~low risk. This is a nice-to-have for achieving free speedups in CI and during ordinary development, but not strictly necessary. It also is likely on track to stabilize in the short term.~~ Resolved by https://github.com/rust-lang/cargo/pull/11224 .
- [ ] [`asm_const`](https://github.com/rust-lang/rust/issues/93332): medium risk. The transitive blocker for this, `inline_consts`, is blocked on a relatively minor syntactic question. However, even this is enough to delay stabilization indefinitely. Fortunately, if syntax is truly the only remaining blocker for `inline_const`, then we may be able to remove it as a blocker for `asm_const`. The path forward here is to either resolve the syntactic question for `inline_const` or to convince Amanieu to decouple it from `asm_const`, which will require getting assurance from those behind the const machinery that the semantics are set in stone.
- [x] [`asm_sym`](https://github.com/rust-lang/rust/issues/93333): ~~medium risk. Blocked on rewriting sym support for `global_asm!`, which is underway (https://github.com/rust-lang/rust/pull/94468) but we need to clarify whether or not that alone will suffice to unblock `asm_sym`. Even if so, it's unclear when that work will be completed.~~ Resolved by https://github.com/enarx/enarx/pull/2303 .
- [x] [`naked_functions`](https://github.com/rust-lang/rust/issues/90957): ~~low risk. The lang team has all but greenlit the stabilization of this feature, contingent on the resolution of a small number of open questions that should not be difficult to resolve. We are actively contributing towards stabilization here: https://github.com/enarx/enarx/issues/1269~~ Resolved by https://github.com/enarx/enarx/pull/2350 .
- [x] Tier 2 `x86_64-unknown-none` (as a way to avoid the need for `-Z build-std`): ~~low risk. This has a prominent advocate in Josh Triplett, and we can help him out by doing the labor of initiating the process. We are actively contributing towards stabilization here: https://github.com/enarx/enarx/issues/1528~~ Resolved by https://github.com/rust-lang/rust/pull/95705 .
- [x] `generic_const_exprs` and `const_mut_refs`: ~~low risk. The code making use of these features needs to be rewritten for SGX2, which is a deliverable for Enarx 0.4, so we just need to ensure that the eventual rewrite doesn't make use of these.~~ Resolved by https://github.com/enarx/enarx/pull/1655 .
- [ ] [`c_size_t`](https://github.com/rust-lang/rust/issues/88345): low risk. These features provide trivial nice-to-have type aliases. I foresee little resistance to stabilizing them, but even if they are not stabilized we could trivially remove our usage of them without difficulty.
- [x] [`core_ffi_c`](https://github.com/rust-lang/rust/issues/94501): ~~low risk. Same as above.~~ Resolved by https://github.com/rust-lang/rust/pull/98315 .
- [ ] [`nonnull_slice_from_raw_parts`](https://github.com/rust-lang/rust/issues/71941): low risk. This is a minor library addition that is already on the verge of stabilization. All that is required is for a few more libs team members to check their boxes and for a stabilization report to be filed, and the latter is something that we can contribute to move this along.
- [ ] [`slice_ptr_len`](https://github.com/rust-lang/rust/issues/71146): low risk. At first glance this appears problematic as concerns have been raised about the feature in general, however this feature provides APIs on three types, and the outstanding concerns only apply to two of those types, whereas the third type is fine, and we're exclusively using this API on that third type. Other users have already noticed this and have split out the non-problematic API into its own feature, which is well on its way towards stabillization (https://github.com/rust-lang/rust/pull/94640) and just needs a nudge to push it over the finish line, which we can aid with.
- [ ] [`slice_ptr_get`](https://github.com/rust-lang/rust/issues/74265): medium risk. This API is currently being employed as a way to avoid potential UB in one place in sallyport (see #2069). There may be other ways to work around the issue, although given the security implications I would prefer to use the standard API if able. Stabilization of this API is highly motivated by it being possibly the only safe way to perform certain obscure unsafe code patterns, but whether or not there is appetite to stabilize it quickly is unknown. It's also possible that the language itself will change to make the use of this API unnecessary.

**Comment by bstrie** [comment]
For a first step we can focus on getting Enarx onto stable, although we'll eventually need to get all of its dependencies onto stable as well (after a cursory examination I think the only dependency currently making use of unstable features is Sallyport).

Here's a list of all the unstable features currently in use by Enarx. These should all eventually have their own tracking issue once some cursory investigation has been done to determine the best course of action for each feature.

* `naked_functions`: https://github.com/enarx/enarx/issues/1269
* `asm_const`
* `asm_sym`
* `generic_const_exprs`
* `const_mut_refs`

**Comment by jarkkojs** [comment]
For me this is a **mandatory requirement** to include Enarx of my regular kernel testing flow. Not having this makes it extremely hard to compile Enarx into initramfs bundled with the kernel image, when compiling a minimal rootfs with BuildRoot. I already have in-progress BuildRoot fork waiting for this: https://github.com/jarkkojs/buildroot-enarx.

**Comment by jarkkojs** [comment]
Looking at your "stable rust" issue, could we take the following "soft path" towards it:
1. If a patch is doable with stable rust with reasonable steps, even if slightly less elegant, then that can be used as an argument in a code review.
2. If a patch is non-doable with stable rust, or requires unreasonable amount of work, then we use unstable rust.
This way we converge to the goal but it does not require any major rework at once.

I.e. we consider stable Rust implementation "better" in a code review when it is possible.

**Comment by jarkkojs** [comment]
> For a first step we can focus on getting Enarx onto stable, although we'll eventually need to get all of its dependencies onto stable as well (after a cursory examination I think the only dependency currently making use of unstable features is Sallyport).
> 
> Here's a list of all the unstable features currently in use by Enarx. These should all eventually have their own tracking issue once some cursory investigation has been done to determine the best course of action for each feature.
> 
> * `naked_functions`: [[Tracking]: Pursue stabilization for Rust RFC 2972: "Constrained Naked Functions" #1269](https://github.com/enarx/enarx/issues/1269)
> * `asm_const`
> * `asm_sym`
> * `generic_const_exprs`
> * `const_mut_refs`

Can you provide some short plain English descriptions what these are?

**Comment by npmccallum** [comment]
@jarkkojs  can you use `RUSTC_BOOTSTRAP` for your kernel flow?

**Comment by jarkkojs** [comment]
> @jarkkojs can you use `RUSTC_BOOTSTRAP` for your kernel flow?

What is it?

**Comment by jarkkojs** [comment]
> One more thing to note for the future: as soon as [rust-lang/cargo#9992](https://github.com/rust-lang/cargo/pull/9992) lands we'll begin using a new, unstable `-Z bindeps` Cargo flag, tracked at [rust-lang/cargo#9096](https://github.com/rust-lang/cargo/issues/9096) . It's too soon to say how long it will take to stabilize once implemented, but I don't think it will take an extremely long time to stabilize.

I fully understand this. Not having that causes some awkwardness in the current codebase.

**Comment by bstrie** [comment]
RUSTC_BOOTSTRAP is a secret implementation detail of the Rust compiler that allows nightly Rust to pretend to be stable Rust, which is used for the purpose of bootstrapping the standard library during the release process. It is not to be used by anyone other than the Rust compiler in any production capacity; if I were a code auditor I would reject any codebase that used it.

**Comment by bstrie** [comment]
Regarding `-Z bindeps`, one of our prerequisites for using artifact dependencies is to port the shims to `x86_64-unknown-none` (#1461, #1462), and that target itself is essentially unstable as it currently requires users to use the unstable `-Z build-std` Cargo flag. Thus compiling with stable Rust will require elevating that target to Tier 2 ( https://doc.rust-lang.org/nightly/rustc/platform-support.html#tier-2 ); this currently has no tracking issue in the Rust repo but I'll ping some people about it.

**Comment by bstrie** [comment]
We may also begin relying on a new feature, `core_ffi_c`. See the rationale at https://github.com/enarx/enarx/issues/1462#issuecomment-1058101996 .

**Comment by bstrie** [comment]
I have confirmed that the proposed https://github.com/enarx/enarx/pull/1608 removes all uses of the `generic_const_exprs` and `const_mut_refs` feature flags.

**Comment by jarkkojs** [comment]
> I have confirmed that the proposed #1608 removes all uses of the `generic_const_exprs` and `const_mut_refs` feature flags.

Not using them. Heap is created dynamically.It allows to manager virtual memory areas with their real addresses. @npmccallum i.e. I'm using ledger with real coordinates now.

**Comment by bstrie** [comment]
I've also just noticed that one of the tests has `#![cfg_attr(target_os = "wasi", feature(wasi_ext))]`. I'll inquire as to the status of stabilization for the wasi_ext feature, but for now this is only a minor concern since it's just a test.

**Comment by bstrie** [comment]
While poking around in our dependencies it appears as though the only external dependency (i.e. not in the Enarx repo) that makes use of unstable features is sallyport. I've identified two new unstable features used by sallyport, `nonnull_slice_from_raw_parts` and `slice_ptr_len`. Both of these features appear to be low risk. I've updated the checklist in the main issue to add these.
